### PR TITLE
Uplift Terra to use React 16

### PR DESCRIPTION
**Pull request by JakeLaCombe**
### Summary
The Pull Request you have all been waiting for! I was looking through the different packages, and here are the biggest non passive changes that affected us

- React 16 removes [data-reactroot] from the React Element
- Enzyme with react-16 does not work with rendering portals. Given our test coverage with nightwatch and wdio covered these cases, I opted just to remove the render tests.
- The new datepicker opts to use popper instead of tether for it's implementation.
- The focusable attribute on SVG nodes takes a string as opposed to a boolean for it's value.
- For our portal usage, we can use the new api as opposed to the legacy one, as indicated by the new imports.
- React 16 requires requestAnimationFrame, meaning we need to add the raf polyfill to work with legacy IE browsers.
- HTML Attributes are more strict. If you had a boolean and it expects a string, it won't convert it

Deployment LInk: https://terra-core-deployed-pr-1194.herokuapp.com/

### CODE REVIEW

**Comment by JakeLaCombe** [comment]
I'll leave commentary on my code review changes so that it provides more context.

File: packages/terra-alert/package.json [hunk 1]
```diff
@@ -21,12 +21,12 @@
   },
   "homepage": "https://github.com/cerner/terra-core#readme",
   "devDependencies": {
-    "react-intl": "^2.3.0",
-    "terra-props-table": "^1.11.0"
+    "react-intl": "^2.4.0",
+    "terra-props-table": "^1.10.1"
```
Comment (JakeLaCombe) [inline packages/terra-alert/package.json]: Bad Merge on my part :(. I'll fix it!

File: packages/terra-alert/tests/jest/Alert.test.jsx [hunk 2]
```diff
@@ -24,7 +24,7 @@ describe('Alert with no props', () => {
 
   it('throws error on missing locale prop in Base', () => {
     try {
-      shallow(<Alert />);
+      render(<Alert />);
```
Comment (JakeLaCombe) [inline packages/terra-alert/tests/jest/Alert.test.jsx]: shallow doesn't pass a context to the component anymore, so we need to do a full render to test that condition.

File: packages/terra-base/src/Base.jsx [hunk 3]
```diff
@@ -48,13 +48,21 @@ class Base extends React.Component {
 
   componentDidMount() {
     if (this.props.locale !== undefined) {
-      i18nLoader(this.props.locale, this.setState, this);
+      try {
+        i18nLoader(this.props.locale, this.setState, this);
```
Comment (JakeLaCombe) [inline packages/terra-base/src/Base.jsx]: For React 16, it no longer renders the component if it errors out. https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html

We can use componentDidCatch, but it gets hard with this component, as we never defined what it should do in that case. We just let it error out previously, which no longer works for us as nothing will render. We can switch to the ErrorComponent when we have established what we would like to do when a locale isn't found.

File: packages/terra-arrange/tests/jest/Arrange.test.jsx [hunk 4]
```diff
@@ -56,25 +56,13 @@ describe('Arrange', () => {
...
     const arrange = <Arrange fitStart={fitStart} fitEnd={fitEnd} fill={fill} align="center" fitStartAttributes={{ style: { maxWidth: '10px' } }} fillAttributes={{ style: { maxWidth: '20px' } }} fitEndAttributes={{ style: { maxWidth: '30px' } }} />;
     const wrapper = shallow(arrange);
-    expect(wrapper.instance().props.fitStart.type).toEqual('img');
-    expect(wrapper.instance().props.fitEnd.type).toEqual('a');
-    expect(wrapper.instance().props.fill.type).toEqual('div');
-    expect(wrapper.instance().props.align).toEqual('center');
-    expect(wrapper.instance().props.fitStartAttributes).toEqual({ style: { maxWidth: '10px' } });
-    expect(wrapper.instance().props.fillAttributes).toEqual({ style: { maxWidth: '20px' } });
-    expect(wrapper.instance().props.fitEndAttributes).toEqual({ style: { maxWidth: '30px' } });
+    expect(wrapper).toMatchSnapshot();
```
Comment (bjankord) [inline packages/terra-arrange/tests/jest/Arrange.test.jsx]: Smart 👍

File: packages/terra-date-picker/src/_PopperContainer.scss [hunk 5]
```diff
@@ -0,0 +1,7 @@
+:local {
+  .popper-container {
+    position: absolute;
+    top: 0;
+    z-index: 2147483647;
```
Comment (bjankord) [inline packages/terra-date-picker/src/_PopperContainer.scss]: Is the intent to set this to a value so high, z-index values on other elements wouldn't be higher?

Comment (JakeLaCombe) [reply packages/terra-date-picker/src/_PopperContainer.scss]: Yeah. It's the same z-index that tether originally set so that it's widgets would display on top.

File: packages/terra-date-time-picker/src/DateTimeUtils.js [hunk 6]
```diff
@@ -40,7 +40,7 @@ class DateTimeUtils {
   }
 
   static formatMomentDateTime(momentDate, format) {
-    return momentDate && momentDate.isValid() ? momentDate.format(format) : '';
+    return momentDate && momentDate.isValid() ? momentDate.format(format) : undefined;
```
Comment (JakeLaCombe) [inline packages/terra-date-time-picker/src/DateTimeUtils.js]: This change is needed to resolve the following error.

```
warning.js:33 Warning: Failed prop type: Invalid prop `selected` of type `string` supplied to `DatePicker`, expected `object`.

```

File: packages/terra-modal/tests/jest/Modal.test.jsx [hunk 7]
```diff
@@ -10,8 +10,3 @@ it('should mount an open modal', () => {
   const modal = mount(<ModalExample />);
   expect(modal).toMatchSnapshot();
 });
-
-it('should render an open modal', () => {
-  const modal = render(<ModalExample />);
```
Comment (JakeLaCombe) [inline packages/terra-modal/tests/jest/Modal.test.jsx]: For those curious, this line produced the following error.

```
<--- Last few GCs --->

[42935:0x102801600]    25833 ms: Mark-sweep 1405.6 (1452.6) -> 1405.6 (1452.6) MB, 1810.7 / 0.0 ms  allocation failure GC in old space requested
[42935:0x102801600]    27430 ms: Mark-sweep 1405.6 (1452.6) -> 1405.6 (1440.6) MB, 1596.5 / 0.0 ms  last resort GC in old space requested
[42935:0x102801600]    29223 ms: Mark-sweep 1405.6 (1440.6) -> 1405.6 (1440.6) MB, 1792.8 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x123410ba5ee1 <JSObject>
    1: read [/Users/jl021230/Documents/PopulationHealth/terra/terra-core/node_modules/react-dom/cjs/react-dom-server.node.development.js:~2206] [pc=0x26d1330d7379](this=0x12341a902ef9 <ReactDOMServerRenderer map = 0x12342cf40529>,bytes=0x1234906829a9 <Number inf>)
    3: renderToStaticMarkup [/Users/jl021230/Documents/PopulationHealth/terra/terra-core/node_modules/react-dom/cjs/react-dom-server....

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/Users/jl021230/.nvm/versions/node/v8.9.2/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/Users/jl021230/.nvm/versions/node/v8.9.2/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/Users/jl021230/.nvm/versions/node/v8.9.2/bin/node]
 4: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/Users/jl021230/.nvm/versions/node/v8.9.2/bin/node]
 5: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/jl021230/.nvm/versions/node/v8.9.2/bin/node]
 6: 0x26d132f8463d
 7: 0x26d1330d7379
```

I posted on enzymes repo, but didn't get much luck in regards to a solution.

File: packages/terra-site/src/TestContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import React from 'react';
+import PropTypes from 'prop-types';
+import styles from './site.scss';
+
+const TestContainer = ({ children }) => (
+  <div className={styles.app} style={{ overflow: 'auto' }}>
+    <div id="wdio-test-wrapper">
```
Comment (JakeLaCombe) [inline packages/terra-site/src/TestContainer.jsx]: This will be our base component. Not though that we still have to have individual for some tests, as they want to display in an individual component inside the wrapper.

File: packages/terra-site/src/examples/embedded-content-consumer/consumers/CustomEventsConsumer.jsx [hunk 9]
```diff
@@ -21,7 +21,6 @@ const CustomEventsConsumer = () => (
     <EmbeddedContentConsumer
       src="#/tests/embedded-content-consumer-tests/custom-events-provider"
       eventHandlers={eventHandlers}
-      fill
```
Comment (JakeLaCombe) [inline packages/terra-site/src/examples/embedded-content-consumer/consumers/CustomEventsConsumer.jsx]: Did we have a reason for these props? I couldn't find any documentation on them, and I found leaving them off still resulted in passing tests, and visual regression didn't reveal any noticeable behavior.

File: packages/terra-status/tests/jest/Status.test.jsx [hunk 10]
```diff
@@ -34,15 +34,6 @@ it('should render a arrange with status and customProps', () => {
   expect(wrapper).toMatchSnapshot();
 });
 
-// Prop Tests
-it('should have all props including customProps set correctly', () => {
```
Comment (JakeLaCombe) [inline packages/terra-status/tests/jest/Status.test.jsx]: This test is covered by the test above. I would have switched this to a snapshot, but the tests on line 34 already does that for us.

File: packages/terra-tabs/src/_TabMenu.jsx [hunk 11]
```diff
@@ -76,6 +76,8 @@ class TabMenu extends React.Component {
 
   wrapOnClick(child) {
     return (event) => {
+      event.stopPropagation();
```
Comment (JakeLaCombe) [inline packages/terra-tabs/src/_TabMenu.jsx]: Essentially, when we click on a tab, the new bounding rectangle is bubbling the click up to the menu click as well, which keeps the menu bar open.

File: packages/terra-tabs/tests/wdio/tabs-spec.js [hunk 12]
```diff
@@ -15,26 +15,26 @@ describe('Tabs - Responsive', () => {
...
     });
     describe('Extended', () => {
       beforeEach(() => {
         browser.url('/#/tests/tabs-tests/extended');
         browser.setViewportSize(viewport);
         browser.moveToObject('.tabContent');
       });
-      Terra.should.matchScreenshot({ viewports });
-      Terra.should.beAccessible({ viewports });
+      Terra.should.matchScreenshot({ viewports: [viewport] });
```
Comment (JakeLaCombe) [inline packages/terra-tabs/tests/wdio/tabs-spec.js]: this should reduce test time. We are already looping through all the viewports at this test, so I don't see a large benefit in having our screenshot comparison match at all of them for every single viewport.

File: packages/terra-time-input/tests/nightwatch/TimeInputTests.jsx [hunk 13]
```diff
@@ -14,8 +14,6 @@ const TimeInputTests = () => (
       <li><Link to="/tests/time-input-tests/twelve-hour-filled-morning">TimeInput - Twelve Hour Default Time Morning</Link></li>
       <li><Link to="/tests/time-input-tests/twelve-hour-filled-evening">TimeInput - Twelve Hour Default Time Evening</Link></li>
       <li><Link to="/tests/time-input-tests/twelve-hour-mobile">TimeInput - Twelve Hour Mobile</Link></li>
```
Comment (JakeLaCombe) [inline packages/terra-time-input/tests/nightwatch/TimeInputTests.jsx]: I had to combine all the tests into the same component due to race conditions. I noticed that sometimes, componentWillMount would render before componentWillUnmount of the previous component, meaning sometimes the fake mobile mode would be disabled. Thus, I kept it at one component on the same page to ensure this doesn't happen.

File: packages/terra-button/tests/wdio/button-spec.js [hunk 14]
```diff
@@ -1,14 +1,4 @@
 /* global browser, Terra, before */
-const shouldTheme = (customProperties, selector) => {
```
Comment (emilyrohrbough) [inline packages/terra-button/tests/wdio/button-spec.js]: Nice. Glad this is finally gone :)

File: packages/terra-date-picker/src/_PopperContainer.scss (hunk 5 above)
Comment (emilyrohrbough) [reply packages/terra-date-picker/src/_PopperContainer.scss]: > It's the same z-index

We should add an inline comment mentioning that

File: packages/terra-icon/src/IconBase.jsx [hunk 15]
```diff
@@ -37,7 +37,7 @@ const propTypes = {
   /**
    * Focusable attribute. IE 10/11 are focusable without this attribute.
    */
-  focusable: PropTypes.bool,
+  focusable: PropTypes.string,
```
Comment (emilyrohrbough) [inline packages/terra-icon/src/IconBase.jsx]: Why this change?

File: packages/terra-icon/CHANGELOG.md [hunk 16]
```diff
@@ -3,6 +3,8 @@ Changelog
 
 Unreleased
 ----------
+### Changed
+* Update to React 16
```
Comment (emilyrohrbough) [inline packages/terra-icon/CHANGELOG.md]: This needs update to reflect the prop type change

File: packages/terra-legacy-theme/CHANGELOG.md [hunk 17]
```diff
@@ -3,6 +3,7 @@ Changelog
 
 Unreleased
 ----------
+* Update to React 16
```
Comment (emilyrohrbough) [inline packages/terra-legacy-theme/CHANGELOG.md]: Not needed

File: packages/terra-markdown/CHANGELOG.md [hunk 18]
```diff
@@ -3,6 +3,8 @@ Changelog
 
 Unreleased
 ----------
+### Changed
```
Comment (emilyrohrbough) [inline packages/terra-markdown/CHANGELOG.md]: Not needed

File: packages/terra-mixins/CHANGELOG.md [hunk 19]
```diff
@@ -3,6 +3,7 @@ Changelog
 
 Unreleased
 ----------
+* Update to React 16
```
Comment (emilyrohrbough) [inline packages/terra-mixins/CHANGELOG.md]: Not needed

File: packages/terra-button/tests/wdio/button-spec.js (hunk 14 above)
Comment (JakeLaCombe) [reply packages/terra-button/tests/wdio/button-spec.js]: I'll get rid of the terra-form-radio on another pull request

File: packages/terra-icon/src/IconBase.jsx (hunk 15 above)
Comment (JakeLaCombe) [reply packages/terra-icon/src/IconBase.jsx]: focusable is actually a string node for SVGs (https://allyjs.io/tutorials/focusing-in-svg.html) and react was throwing prop type errors when we gave it a boolean instead of a string.

**Comment by JakeLaCombe** [comment]
terra-form-radio has it's own implementation of wdio, but to get this pull request moving, I left off that change to another future pull request. It will produce tons of snapshots.

File: packages/terra-tabs/tests/wdio/tabs-spec.js [hunk 20]
```diff
@@ -15,26 +15,26 @@ describe('Tabs - Responsive', () => {
...
```
Comment (emilyrohrbough) [inline packages/terra-tabs/tests/wdio/tabs-spec.js]: You should be able reduce this to Terra.should.matchScreenshot();

File: packages/terra-tabs/src/_TabMenu.jsx (hunk 11 above)
Comment (emilyrohrbough) [reply packages/terra-tabs/src/_TabMenu.jsx]: Is this something you found with the wdio screenshots?? Or in other test instances?

File: packages/terra-button/tests/wdio/button-spec.js (hunk 14 above)
Comment (emilyrohrbough) [reply packages/terra-button/tests/wdio/button-spec.js]: Actually that one is different and needs to stay for now. The toolkit implementation will apply each variable one at at time and take a screenshot for each change. To get the correct theming screenshots for form-radio, multiple variables need to change at one time due to the default being browser naive.

File: packages/terra-tabs/tests/wdio/tabs-spec.js (hunk 20 above)
Comment (JakeLaCombe) [reply packages/terra-tabs/tests/wdio/tabs-spec.js]: Thanks! Removed it, and it works great!

https://github.com/cerner/terra-core/pull/1194/commits/799ad0e50a2c0280d130342f43be767f9d020549

File: packages/terra-tabs/src/_TabMenu.jsx (hunk 11 above)
Comment (JakeLaCombe) [reply packages/terra-tabs/src/_TabMenu.jsx]: I found it through the wdio screenshots, but then tested it in a browser and saw the same behavior.

**Comment by emilyrohrbough** [comment]
@JakeLaCombe Do you have a documented list of React16 changes we should be aware of to for future development?

**Comment by JakeLaCombe** [comment]
@emilyrohrbough What I have listed in the pull request description are so far the main ones I have stumbled upon. I think the only real thing to add would be potentially to ensure that your wdio tests have a base wrapper or selector to choose from.

**Comment by mmalaker** [comment]
Found issue in RtL orientation, when you click on the calendar button for a date picker within a modal, the calendar does not display. I'm testing the rest of the site in RTL to see if there are any other bugs.

**Comment by mmalaker** [comment]
Rechecked the RTL date picker issue and that is now resolved. No other issues found. Functional verification completed.

**Comment by noahbenham** [comment]
I know this is closed but we were running into similar heap memory issues with our project. If this continues happening for `terra-core` and wasn't isolated to that line you removed, here's what we did @JakeLaCombe.

Root cause was including `babel-polyfill` in our `jestSetup.js`. Removing that took our heap memory usage from ~1.5GB to ~125MB. It doesn't look like that's true for this repo though.

We diagnosed our issue with `node --expose-gc ./node_modules/.bin/jest --runInBand --logHeapUsage`. That might be useful in case this continues happening and wasn't caused by the line you removed.
